File: backend/app/api/v1/moves.py
from typing import List
from uuid import UUID
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from app.api.deps import get_db
from app.schemas.move import Move, MoveCreate, MoveUpdate
from app.services.move_service import MoveService

router = APIRouter()

@router.post("/", response_model=Move, status_code=status.HTTP_201_CREATED)
def create_move(
    move_data: MoveCreate,
    db: Session = Depends(get_db)
):
    """Create a new Move for a Pokemon"""
    return MoveService.create_move(db, move_data)

@router.get("/pokemon/{pokemon_id}", response_model=List[Move])
def get_moves_by_pokemon(
    pokemon_id: UUID,
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db)
):
    """Get all moves for a specific Pokemon"""
    return MoveService.get_moves_by_pokemon(db, pokemon_id, skip=skip, limit=limit)

@router.get("/{move_id}", response_model=Move)
def get_move(
    move_id: UUID,
    db: Session = Depends(get_db)
):
    """Get a Move by ID"""
    return MoveService.get_move(db, move_id)

@router.put("/{move_id}", response_model=Move)
def update_move(
    move_id: UUID,
    move_update: MoveUpdate,
    db: Session = Depends(get_db)
):
    """Update a Move"""
    return MoveService.update_move(db, move_id, move_update)

@router.delete("/{move_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_move(
    move_id: UUID,
    db: Session = Depends(get_db)
):
    """Delete a Move"""
    MoveService.delete_move(db, move_id)
    return None

@router.post("/{move_id}/complete", response_model=Move)
def complete_move(
    move_id: UUID,
    db: Session = Depends(get_db)
):
    """Mark a move as completed (execute the task)"""
    return MoveService.complete_move(db, move_id)

@router.get("/pokemon/{pokemon_id}/completed", response_model=List[Move])
def get_completed_moves(
    pokemon_id: UUID,
    db: Session = Depends(get_db)
):
    """Get all completed moves for a Pokemon"""
    return MoveService.get_completed_moves(db, pokemon_id)

@router.get("/pokemon/{pokemon_id}/pending", response_model=List[Move])
def get_pending_moves(
    pokemon_id: UUID,
    db: Session = Depends(get_db)
):
    """Get all pending (incomplete) moves for a Pokemon"""
    return MoveService.get_pending_moves(db, pokemon_id)

File: frontend/src/api/moves.ts
import { apiClient } from './client';
import { Move } from '../types/move';

export const movesApi = {
  // Create move
  create: async (data: { pokemon_id: string; name: string; description?: string; power: number }): Promise<Move> => {
    const response = await apiClient.post('/moves', data);
    return response.data;
  },

  // Get moves by pokemon
  getByPokemon: async (pokemonId: string): Promise<Move[]> => {
    const response = await apiClient.get(`/moves/pokemon/${pokemonId}`);
    return response.data;
  },

  // Get single move
  getById: async (id: string): Promise<Move> => {
    const response = await apiClient.get(`/moves/${id}`);
    return response.data;
  },

  // Update move
  update: async (id: string, data: Partial<{ name: string; description?: string; power: number }>): Promise<Move> => {
    const response = await apiClient.put(`/moves/${id}`, data);
    return response.data;
  },

  // Delete move
  delete: async (id: string): Promise<void> => {
    await apiClient.delete(`/moves/${id}`);
  },

  // Complete move (mark as done)
  complete: async (id: string): Promise<Move> => {
    const response = await apiClient.post(`/moves/${id}/complete`);
    return response.data;
  },

  // Get completed moves for a pokemon
  getCompleted: async (pokemonId: string): Promise<Move[]> => {
    const response = await apiClient.get(`/moves/pokemon/${pokemonId}/completed`);
    return response.data;
  },

  // Get pending moves for a pokemon
  getPending: async (pokemonId: string): Promise<Move[]> => {
    const response = await apiClient.get(`/moves/pokemon/${pokemonId}/pending`);
    return response.data;
  },
};